Service manager v4

ramchale opened the pull request:
This still needs a lot of tidy up, but seems fairly close to a working version with service-manager v4.

Raised as a draft to check if there's anything else obvious I'm missing, and seek some help fixing the remaining tests.

(The block of commented out code in Factory being an obvious area that needs a look)

ramchale committed:
WIP

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
WIP

Signed-off-by: ramchale <legionsofbob@gmail.com>

gsteel commented:
Just playing around locally with this patch - and validator ^3.6 will defo be needed so that we can `$validators->build(ValidatorChain::class, $spec)`.

I think we might need to do some more cleanup before we can figure out what's tripping up those tests… 🤔

ramchale commented:
Thanks for the work on resolve the underlying changes @gsteel

I've just done a quick rebase and fixed a few obvious bits. Busy day ahead, so I'll get back to this when I can. Happy for anyone else to make progress with it in the meantime if they'd like to

gsteel commented:
Cheers @ramchale 👍

gsteel committed:
Clean up `InputFilterFactory`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Clean up `InputFilterPluginManager`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Bump laminas-validator to compatible version

Signed-off-by: George Steel <george@net-glue.co.uk>

ramchale committed:
Looking into test failures

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale commented on test/InputFilterAbstractServiceFactoryTest.php:
```diff
@@ -54,78 +55,82 @@ protected function setUp(): void
+        self::assertSame($validator, $validatorPluginManager->get('foo_validator'));
+        // End
+
+        $inputFilter = $this->factory->__invoke($services, 'filter');
```
This fails with "ServiceNotFoundException: Unable to resolve service "foo_validator" to a factory"

Debugging thorugh, it looks like the filter also can't be resolved but does not throw an Exception.
The build methods of the filterPluginManager and validatorPluginManager don't seem like they try to use the 'services', so I'm not sure this can pass unless we want to make further changes to those.

Worth double checking though, as I'm much less familiar with the internals of the plugin managers than some other parts of the code

gsteel replied:
> The build methods of the filterPluginManager and validatorPluginManager don't seem like they try to use the 'services',

Yes! This is what had me a bit stumped… `build()` will never consult `services`, only `get()` will.

I've added fixes to both `filter` and `validator` which should help us move forward:

- https://github.com/laminas/laminas-filter/pull/261
- https://github.com/laminas/laminas-validator/pull/435

ramchale replied:
Thanks @gsteel
Looks like it gets over that hurdle 😄

gsteel replied:
Tests should pass once https://github.com/laminas/laminas-validator/pull/437 is merged.

There are a number of skipped tests that focus on runtime validation of validator/filter spec arrays - runtime validation of these is gone, so, we need to figure out what to do there…

ramchale replied:
@gsteel Been a while as I've got some life stuff going on, but yep the tests are looking a lot better. Thanks.
Is the intent for things like FactoryTest->testCreateInputWithInvalidFilterSpecType that static analysis is relied on to pick up these, so we can remove the test? At least in the cases where we've needed to add `/** @psalm-suppress InvalidArgument */`

gsteel replied:
Hey @ramchale 👋

I've been bitten recently removing defensive runtime checks in favour of static analysis. I think you have to look at that sort of stuff on a case-by-case basis.

For things like array spec - you have to know the spec up-front by reading docs, so, if you get that wrong as a user, I think it's OK to have `TypeError` rather than defensive runtime checks, but also, it's easy to make mistakes with array specs and clear exception messages really help there.

I hope the life stuff is _good_ life stuff 👍

ramchale replied:
We can add some type checking in input-filter as in the [latest commit ](https://github.com/laminas/laminas-inputfilter/pull/149/commits/18b1fc7925aba0cc39a8287ea242bd73a4162a2b), but a couple of the skipped tests feel like the type validation more naturally sit in laminas-filter and laminas-validator, closer to where the original type error the test looks for originated. I can put in PRs to those repositories add checks for the presence of the required array keys with more helpful error messages if that works for you?

ramchale committed:
Quick check with dev branches

Signed-off-by: ramchale <legionsofbob@gmail.com>

gsteel committed:
Bump validator and filter to most recent patch release

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
CS Fixes

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Build chains via plugin managers

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fixes some and skips other tests…

Todo: Tests expecting runtime validation of specification arrays all now fail because upstream libs rely on static analysis rather than runtime validation.

Either we need to change the libs to perform runtime validation, or, implement it here, or, let the errors/warnings fly.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Misc WIP on tests

Signed-off-by: George Steel <george@net-glue.co.uk>

ramchale committed:
Update laminas/laminas-validator

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Specify isValid context and return type

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Remove unused FactoryTest->createInputFilterPluginManagerMockForPlugin

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Add type check for filters specification

Signed-off-by: ramchale <legionsofbob@gmail.com>

gsteel committed:
Introduce configuration constant

Adding the constant and constructor means we don't have to define values for deprecated properties

Also removes legacy zend* aliases

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Suppress Psalm Issues in SA Test Code

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve types for data provider

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
"Un-Skip" currently failing tests

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
QA: work around calls to deprecated methods and constants

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Reset baseline

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Rebase onto 3.0.x, Refresh locked dependencies

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Make test adjustments after rebase onto 3.0.x

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Reinstate upgrade to php 8.2 - 8.5

Signed-off-by: George Steel <george@net-glue.co.uk>

ramchale committed:
Tweak PHP requirement

Signed-off-by: ramchale <legionsofbob@gmail.com>

gsteel committed:
Validate Filter & Validator Specs

Upgrades filter and validator to versions containing runtime validation tools for the various array shapes.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Merge remote-tracking branch 'ramchale/service-manager-v4' into service-manager-v4

Signed-off-by: George Steel <george@net-glue.co.uk>

# Conflicts:
#	composer.json
#	composer.lock

gsteel commented:
@ramchale We've got green CI 🎉

Is there anything else to do in this patch or shall we review and get it merged?

gsteel commented on src/Factory.php:
```diff
@@ -88,39 +90,37 @@ private function canCreateInputType(string $class): bool
+        $validators = $spec['validators'] ?? [];
+
+        if (! is_array($filters) && ! is_callable($filters) && ! $filters instanceof FilterInterface) {
+            throw new TypeError("filters must be an array, callable, or FilterInterface. Received: "
```
This should not be a `TypeError`, it should be `InvalidArgumentException` in this NS

ramchale replied:
Happy to change it, but currently testCreateInputWithInvalidFilterSpecType is testing for TypeError so it's a slight change to the expected behaviour

gsteel replied:
I think it was me that introduced the `TypeError` changes in this patch and previously it was something else.

gsteel commented on src/Factory.php:
```diff
@@ -88,39 +90,37 @@ private function canCreateInputType(string $class): bool
+     *     validatorChain: ValidatorChainInterface,
      * }
      */
     private function buildChainsFromSpecification(array $spec): array
```
We should probably add all the `@throws` here so it's easier to amend the public API to include `@throws` where necessary

gsteel commented on test/FactoryTest.php:
```diff
@@ -103,23 +97,20 @@ public function testCreateInputWithTypeAsAnInvalidClassInstanceThrowException():
     {
         $factory = $this->createDefaultFactory();
-
         $this->expectException(TypeError::class);
```
This test should be checking for a specific exception type in our NS

ramchale commented:
Awesome work @gsteel !

Aside from squashing the commits, no. I think anything else can be a seperate PR

gsteel commented on test/FactoryTest.php:
```diff
@@ -287,7 +273,10 @@ public function testFactoryWillCreateInputWithSuggestedFilters(): void
+                    self::assertEquals(
+                        'iso-8859-1',
+                        (new ReflectionObject($filter))->getProperty('encoding')->getValue($filter)
+                    );
```
This test needs completely re-working to get rid of reflection introduced because of dropped upstream getters.

For example:

1) Create a filter chain with a predicatable output
2) Assert that executing the filter yields the desired output

We can assert the size of the chain with `assertCount(2, $chain)` because I think it's `Countable`

gsteel commented on test/FactoryTest.php:
```diff
@@ -299,16 +288,16 @@ public function testFactoryWillCreateInputWithSuggestedFilters(): void
     {
         $factory = $this->createDefaultFactory();
-        $digits  = new Validator\Digits();
+        $digits  = new Digits();
```
Same thing with the equivalent validator chain test, make the chain then assert the chain fails or passes validation for given inputs based on validators added and options.

gsteel commented on test/InputFilterAbstractServiceFactoryTest.php:
```diff
@@ -54,102 +59,108 @@ protected function setUp(): void
-        $this->services->setService('config', []);
+        if ($config !== null) {
+            $services->setService('config', $config);
+        }
```
Instead of `setService`, we should inject into the constructor with `new ServiceMananager(['services' => ['config' => $config]]);`

gsteel replied:
Or conditionally inject it when non-null, i.e `$options = $config === null ? [] : [...]`

gsteel commented on test/InputFilterPluginManagerCompatibilityTest.php:
```diff
@@ -38,41 +41,33 @@ protected static function getPluginManager(array $config = []): InputFilterPlugi
 
     public function testShareByDefaultAndSharedByDefault(): void
```
```suggestion
    public function testInputFiltersAreNotSharedByDefault(): void
```

gsteel reviewed (changes requested):
Spotted a few nitpicks

ramchale commented:
> Spotted a few nitpicks

Thanks. I'll pick most of those up tomorrow

ramchale committed:
Throw InvalidArgumentException from buildChainsFromSpecification

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Rename testShareByDefaultAndSharedByDefault to testInputFiltersAreNotSharedByDefault

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Configure ServiceManager through constructor in testCanCreate

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Remove reflection from testFactoryWillCreateInputWithSuggestedFilters

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale committed:
Remove reflection from testFactoryWillCreateInputWithSuggestedValidators

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale commented:
@gsteel think I've covered off all the feedback. Let me know if there's anything else you can see or if it's good to sqash

gsteel reviewed (approved):
Thanks @ramchale LGTM!
